store/store.go

```diff
@@ -159,30 +165,68 @@ func (s *Store) Put(key []byte, value []byte) error {
 		return err
 	}
 
-	has, err := s.Has(key)
+	// Get the key in primary storage
+	indexKey, err := s.index.Primary.IndexKey(key)
 	if err != nil {
 		return err
 	}
-	// we assume that a second put should NOT write twice
-	// the reason is we are assuming the key is the immutable hash for the value
-	if has {
-		return ErrKeyExists
-	}
-	fileOffset, err := s.index.Primary.Put(key, value)
+	// See if the key already exists and get offset
+	prevOffset, found, err := s.index.Get(indexKey)
 	if err != nil {
 		return err
 	}
-	indexKey, err := s.index.Primary.IndexKey(key)
+	// If found get the key and value stored in primary to see if it is the same
+	// (index only stores prefixes)
+	var storedKey []byte
+	var storedVal []byte
+	if found {
+		storedKey, storedVal, err = s.index.Primary.Get(prevOffset)
+		if err != nil {
+			return err
+		}
+	}
+
+	if bytes.Compare(key, storedKey)+bytes.Compare(value, storedVal) == 0 {
+		// We are trying to put the same value in an existing key,
+		// we can directly return
+		// NOTE: How many times is going to happen this. Can we save ourselves
+		// this step? We can't in the case of the blockstore.
+		return nil
+	}
+
+	// We are ready now to start putting/updating the value in the key.
+	// Put value in primary storage first
+	fileOffset, err := s.index.Primary.Put(key, value)
 	if err != nil {
 		return err
 	}
-	if err := s.index.Put(indexKey, fileOffset); err != nil {
+
+	// If the key being set is not found, or the stored key is not equal
+	// (even if same prefix is shared @index), we put the key without updates
+	if !found || bytes.Compare(key, storedKey) != 0 {
+		if err := s.index.Put(indexKey, fileOffset); err != nil {
+			return err
+		}
+	}
+	// If the key exists and the one stored is the one we are trying
+	// to put this is an update.
+	if found && bytes.Compare(key, storedKey) == 0 {
```

Comment:
is this condition the else of the above condition?

Reply:
I believe they're inverses of each other.
let's say `!found`  is `a` and `bytes.Compare(key, storedKey) != 0` is `b`
the the first `if` is  `a || b`
and the second `if` is `!a && !b`
`!a && !b` is the same as `!(a || b)` so the conditions are inverses, and you can just put the update in the else block.

Reply:
*non-blocking*